### DISCUSSION TITLE
API and implementation proposal for BigInt

### DISCUSSION THREAD

**Discussion by hungrybluedev** [vlang/v#9578] [Ideas]
# Update

`big.Integer` has been implemented: https://github.com/vlang/v/pull/11352

---------
# Overview

This is a proposal for the steps to be taken to implement `BigInt` for vlib. The plan is to implement everything in V in order to keep everyone on the same page and support all backends. Feature completeness is a higher priority than performance at the moment. Naive algorithms will be implemented and tested first. Improvements to performance in the form of implementation of better algorithms should be PRs and should not regress existing unit tests.

# State of the Art

Here is a _very brief_ overview of what the current scenario is in established languages:

1. GoLang has a mixture of pure Go and assembly code. [Source](https://github.com/golang/go/tree/master/src/math/big).
2. Julia, Crystal have bindings to GMP. [Julia Docs](https://docs.julialang.org/en/v1/manual/integers-and-floating-point-numbers/#Arbitrary-Precision-Arithmetic) and [Crystal Docs](https://github.com/crystal-lang/crystal/tree/master/src/big).
3. Java does not use bindings. [Source](https://github.com/openjdk/jdk/tree/master/src/java.base/share/classes/java/math).

Currently, V wraps [tiny bignum](https://github.com/kokke/tiny-bignum-c). @hanabi1224 has an [implementation](https://github.com/hanabi1224/vlang-big-integer) of `BigInt` that we aim to re-implement and incorporate into the standard library.

# Proposed API

Here are the proposed functionalities that are a necessity:

1. Arithmetic operators: `+`, `-`, `*`, `/`, `%`, `div_mod`.
2. Comparison operators: `==` and `<`. (The remaining will be auto-generated by the compiler).
3. Conversion methods: string to int(s), int(s) to string, different bases.
4. Bitwise operators.
5. GCD

More functions can be added later when needed.

# References

~Word concatenation implementation: https://hackage.haskell.org/package/adb-0.1.0.0/src/Network/ADB/bigint.c~

This implementation seems to use a regular digit array implementation.

# Tentative Roadmap

- [ ] Implement API
- [ ] Replace existing usages of the `big` module.
- [ ] Deprecate the `big` module _or_ replace carefully with new one.
- [ ] Extensive unit testing and documentation.

# After This

Following ~`BigInt`~ `big.Integer`, the types to be implemented could be:

1. ~`BigFrac`~ `big.Rational`
2. ~`BigDecimal`~ `big.Decimal`
3. ~`BigFloat`~ `big.Float` (optional)

**Comment by penguindark** [comment]
Do we will start with a 4byte per digit implementation or with the concatenation of words?

**Reply by hungrybluedev** [reply]
We'll go with concatenation of words. We will prioritize faster operations over faster conversions.

**Comment by crthpl** [comment]
1. Will `BigFrac` be what some call a rational?
2. What is `BigDecimal`? Two `BigInt`s controlling right and left of decimal point? How would that work

**Reply by hungrybluedev** [reply]
1. Yes it will be a rational number. It would be best to use the names suggested by @spytheman : `big.Integer`, `big.Decimal`, `big.Rational`.
2. No need for two ints. Just one `big.Integer` for the coefficient and one regular int for the exponent. Additionally, we could keep track of the precision. I have some prior experience with this in proprietary codebases.

**Comment by Cons-Cat** [comment]
I once built an idle game that needed numbers to scale very very large. However, being a non-scientific application, it wasn't really important that the math was precise. Iirc, mine was guaranteed to never exceed the size of 8 bytes for a number + 4 bytes to represent exponent (which I think represented multiples of 3 for whatever reason, i.e. exponent = 2 means the number is exponentiated to the 6th). Or something like that. So it wasn't truly Big, but could represent quite large values with an acceptable precision that players wouldn't notice failing. I believe that Cookie Clicker does something similar. Maybe game applications are a niche, but I think letting programmers write something like this ...
``big_int := big.StaticInt<8, 4>{ /**/ }``
... would be useful enough to be worthwhile for the standard library, personally.

**Reply by hungrybluedev** [reply]
It's an interesting idea but I think it is more niche than arbitrary precision arithmetic itself (which is not a matter of concern for a majority of developers).

After a basic implementation of `big.Integer` is complete, you can try suggesting an API, or even better, offering an implementation for `big.StaticInteger`. I'll be available for review.

**Comment by VincentLaisney** [comment]
Hi! I have hit on problems with the bigint implementation of vlib with the Sqrt_by_hand snippet on RosettaCode.
I have made some code for harnessing the big-integer library of vlang. https://github.com/VincentLaisney/bigint_harness 
The actual implementation with tiny-bigint gives erroneous results for:
- all negative result. It is normal (to be expected) since it is unsigned, but handling of negative is necessary for a bigint library
- large number. Until 100 decimal digits it is OK but for 200 each number is faulty. A solution is to change the source of the library to augment the size of the Number struct but in any case it can not handle an arbitrary long number.

I have tried using the hanabi1224.biginteger. It is also not accurate even for small number.
For example, I get:
8337839423 - 9683495887 = -36893488146073446768
instead of -1345656464

If I have some time I shall try to make in interface with a reliable library (like GMP) or port the Go-library which is written completely in Go and readable.

**Reply by hungrybluedev** [reply]
The problem with GMP is the licence. If there's an MIT licensed alternative, we can quickly incorporate it and add it as a third-party dependency.

However, we'd prefer to write a library in pure V. Currently, I'm a bit preoccupied with the V Language Server project but I want to implement big int, big rational, and big decimal, possibly as a part of my thesis. I'll definitely accept your help if you're up for it. You can get in touch with us on the [Discord Server](https://discord.gg/vlang).

**Reply by VincentLaisney** [reply]
Writing a big-number library is very delicate as I have seen in the Go-code that I have read and in the bug I have found in some libraries see my harnessing project  https://github.com/VincentLaisney/bigint_harness.
I am very glad that you are ready to write a library.
I think to make my wrapper an external module, so everybody can use it.

**Reply by hungrybluedev** [reply]
I'd suggest the following:

- [ ] Try to find an MIT big integer library in C to wrap so that there are no licensing problems.
- [ ] Add several unit tests. You can take those from elsewhere, as well as write your own.

Also, keep us updated about your progress.

**Reply by hanabi1224** [reply]
@VincentLaisney  Hey regarding ```8337839423 - 9683495887 = -36893488146073446768```, it's a bug and I just made a fix. That lib was made for fun and does not have enough test coverage, sorry about that.

**Reply by blackshirt** [reply]
Maybe LibBF, from Fabrice Bellard, the man behind Qemu, ffmpeg, tiny C compiler and soon could be a good candidate for this purpose ...here's the link https://bellard.org/libbf/

**Reply by VincentLaisney** [reply]
I have put a first version of porting the GMP to V.
It still lacks the docs and above all it leaks. I have add a .free() method but it is not called by the V code (I have looked in the generated .c file). I shall ask on discord

**Reply by VincentLaisney** [reply]
It is on https://github.com/VincentLaisney/v_gmp/tree/main

**Reply by VincentLaisney** [reply]
@blackshirt. Thank you. I'll try it.

A complementary note. I have tried the v_gmp library against my bigint-harnessing and there is no problem. GMP is very quick.

**Reply by penguindark** [reply]
@VincentLaisney to manage manually memory (autofree is a work in progress) you must indicate to the compiler that your function use the manual memory management usign the flag `[manualfree]`: https://github.com/vlang/v/blob/master/doc/docs.md#control.
Otherwise for the moment you can use the Garbage collector using the flag `-gc boehm` in the compile call: `v -gc boehm file.v`

**Reply by VincentLaisney** [reply]
Thank you @penguindark

**Reply by VincentLaisney** [reply]
When I have added the flag -autofree the free() method of my struct is correctly called. Everything is therefore OK.

**Reply by blackshirt** [reply]
> It is on https://github.com/VincentLaisney/v_gmp/tree/main

I have test it in my code, as a backend for doing calculation in elliptic curve arithmatics, i think its work, even there are some leaks in the detail

**Comment by VincentLaisney** [comment]
I am writing an interface for a Biginteger module with libbf but this library is state-free and for each function a precision and a rounding or flags or a context is necessary.
I want to make an user-friendlier API and put the these parameters as "static" variables in the module. So can I override the operators.
I have tried to put the 'static' as __global's in the module but for compiling I need to add the flag -enable-globals.
I do not want that the clients of the library have to do so.
Is there another solution to have 'static' var in a module? Maybe an Attribute? Or I'll do a C file only for these 'static'.

Thank you for your help.

**Reply by hungrybluedev** [reply]
Personally, I plan to use a `MathContext` for storing rounding and precision information. Similar in spirit to Java's implementation. Then have the struct initialised with sensible defaults. In case a user wants to change some settings, they may pass along the updated context. We use structs in place of named parameters when calling functions.

Btw, you can look at the progress I have made on `big.Integer` [here](https://github.com/hungrybluedev/v/projects/1). If you are up to it, you can help implement a few features.

**Reply by VincentLaisney** [reply]
I can use a struct as well, but how put it as static in the library without (or with) using __global's?
For a Bigint + Bigint = Bigint function we cannot use another parameter.
Thanks for the link.

**Reply by hungrybluedev** [reply]
I understand the inconvenience. In this case, I would suggest keeping a well documented, wordy function like `add(BigInt, Context)` and then overloading the `+` with a default value.

**Reply by VincentLaisney** [reply]
In the LibBF the same functions are used for both Biginteger and Bigfloat. The difference is, for some function, an INFINITE_PRECISION for Bigint.
It is impossible to have the same default Context for both type. Therefore I wanted to have a static context so that the user can fix once the precision (Integer_precision or float_precision).

If there could not be globals, I see two solutions. Change the types adding a field in one struct and duplicating all the functions to have two separate API's. The other is doing the globals at the C-level.
The first solution is cleaner and I would prefer it.

I have already done a part of the work and tests and the distinction between the functions for both type is important. For example 'div' makes a segmentation fault with the infinite precision. For Integer one must use another function 'divrem'.

**Reply by dumblob** [reply]
You could maybe use generics for the "different" contexts...

But globals and "scope-constrained globals" (aka static variables) are a no-go for multithreading-first language like *V*. So please find a different solution.

Some time ago I've proposed *attributes* for module imports to set/change defaults when importing it. This could work, but has the downside that there wouldn't be any easy way to change the behavior after the module got imported. And that's also the reason why I wouldn't want my proposal be implemented any more (or at least not without significant enhancements).

**Comment by VincentLaisney** [comment]
I have nearly finished the interface to libBF (only for BigInt for now). There is still a strange bug only in the subtraction (shown by my bigint-harnessing).
It is very strange especially since subtraction is only changing of the sign and addition.
I'll use gmp to make tests and compare the results with those of libbf.

**Comment by VincentLaisney** [comment]
I have posted a first version of v_libbf of GitHub: https://github.com/VincentLaisney/v_libBF/tree/main
